File: 백준/Gold/14502. 연구소/연구소.java
import java.io.*;
import java.util.*;

public class Main {
    static int[][] graph;
    static int n;
    static int m;
    static int result;

    static Boolean nearWall(int r, int c) {
        int[] dx = {1,-1, 0, 0, 1, 1, -1, -1};
        int[] dy = {0, 0, 1, -1, 1, -1, 1, -1};

        for (int i = 0; i < 8; i++) {
            int nr = r + dx[i];
            int nc = c + dy[i];

            if (nr < 0 || nc < 0 || nr >= n || nc >= m) {
                return true;
            } else if (graph[nr][nc] >= 1) {
                return true;
            }
        }
        return false;
    }

    static int getCount() {
        boolean[][] visited = new boolean[n][m];
        int safe = 0;
        int[] dx = {1, -1, 0, 0};
        int[] dy = {0, 0, 1, -1};

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (visited[i][j] || graph[i][j] != 2) {
                    continue;
                }

                ArrayDeque<int[]> stack = new ArrayDeque<>();
                stack.push(new int[]{i, j});

                while (stack.size() > 0) {
                    int[] cur = stack.pop();
                    int r = cur[0];
                    int c = cur[1];

                    if (visited[r][c]) {
                        continue;
                    } else {
                        visited[r][c] = true;
                    }
                    
                    for (int k = 0; k < 4; k++) {
                        int nr = r + dx[k];
                        int nc = c + dy[k];

                        if (nr < 0 || nc < 0 || nr >= n || nc >= m) {
                            continue;
                        } else if (graph[nr][nc] == 1 || visited[nr][nc]) {
                            continue;
                        } else {
                            stack.push(new int[]{nr, nc});
                        }
                    }
                }
            }
        }

        for (int i = 0; i < n; i++) {
            for (int j =0; j < m; j++) {
                if (graph[i][j] == 0 && !visited[i][j]) {
                    safe += 1;
                }
            }
        }

        return safe;
    }


    static void backtrack(int limit, int cnt) {
        if (cnt == 3) {
            result = Math.max(result, getCount());
            return;
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (graph[i][j] != 0) {
                    continue;
                } else if (limit >= i * n + j) {
                    continue;
                } else if (!nearWall(i,j)) {
                    continue;
                } else {
                    graph[i][j] = 1;
                    backtrack(i * n + j, cnt+1);
                    graph[i][j] = 0;
                }
            }
        }

    }


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        graph = new int[n][m];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                graph[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        backtrack(-1, 0);
        System.out.println(result);
    }
}

File: 백준/Gold/14502. 연구소/연구소.py
# 연구소
import sys
input = sys.stdin.readline

def nearWall(i, j):
    direc = [(1,0), (-1,0), (0,1), (0,-1), (1,1), (1,-1), (-1,1), (-1,-1)]

    for dx, dy in direc:
        nr, nc = i + dx, j + dy
        if nr < 0 or nr >= n or nc < 0 or nc >= m:
            return True
        elif graph[nr][nc] >= 1:
            return True

    return False

def getCount():
    visited = [[0] * m for _ in range(n)]
    safe = 0
    
    for i in range(n):
        for j in range(m):
            if visited[i][j] or graph[i][j] != 2:
                continue

            stack = [(i, j)]
            while stack:
                r, c = stack.pop()

                if visited[r][c]:
                    continue
                else:
                    visited[r][c] = 1
                    
                for dx, dy in [(1,0), (-1,0), (0,1), (0,-1)]:
                    nr, nc = r + dx, c + dy
                    if nr < 0 or nr >= n or nc < 0 or nc >= m:
                        continue
                    elif graph[nr][nc] == 1 or visited[nr][nc]:
                        continue
                    else:
                        stack.append((nr, nc))

    for i in range(n):
        for j in range(m):
            if graph[i][j] == 0 and not visited[i][j]:
                safe += 1

    return safe
            
def backtrack(limit, cnt):
    global result
    
    if cnt == 3:
        result = max(result, getCount())
        return

    for i in range(n):
        for j in range(m):
            if visited[i][j] or graph[i][j] != 0:
                continue
            elif limit >= i * n + j:
                continue
            elif not nearWall(i, j):
                continue
            else:
                visited[i][j] = 1
                graph[i][j] = 1
                backtrack(i * n + j, cnt+1)
                visited[i][j] = 0
                graph[i][j] = 0
    
n, m = map(int,input().split())
graph = []
visited = [[0] * m for _ in range(n)]
result = 0

for i in range(n):
    graph.append(list(map(int,input().split())))

backtrack(-1, 0)

print(result)

